lib/recursive-methods.rb

```diff
@@ -1,49 +1,108 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError
+  end
+  if n == 0
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.empty?
+    return ""
+  end
+  if s.length == 1
+    return s[0]
+  else
+    return s[s.length - 1] + reverse(s[0..s.length - 2])
+  end
+end
+
+# Time complexity: O(n)
+# Space complexity: O(1)
+def reverse!(s, b, e)
+  if b >= e
+    return
+  else
+    temp = s[b]
+    s[b] = s[e]
+    s[e] = temp
+    reverse!(s, b + 1, e - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.empty?
+    return ""
+  end
+  if s.length == 1
+    return s[0]
+  end
+  b = 0
+  e = s.length - 1
+  reverse!(s, b, e)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  if n == 0
+    return 0
+  else
+    return 2 + bunny(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 0
+    return true
+  elsif s.length == 1
+    return false
+  end
+  return s[0] == "(" && s[s.length - 1] == ")" && nested(s[1..s.length - 2])
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def search(array, value)
-    raise NotImplementedError, "Method not implemented"
+  if array.empty?
+    return false
+  end
+  if array[0] == value
+    return true
+  else
+    return search(array[1..array.length - 1], value)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
```

Comment:
O(n^2), think about how you solved reverse in place, could you use that technique to make this more efficient?